### PR TITLE
WebGLRenderer: Support custom painter sort functions.

### PR DESCRIPTION
**Pull request by ranbuch**
As discussed [here](https://github.com/mrdoob/three.js/issues/13889#issuecomment-385225320) I've created a PR that enables support for dynamic sort functions.

### CODE REVIEW

File: src/renderers/webgl/WebGLRenderLists.js
Comment (D504) [inline src/renderers/webgl/WebGLRenderLists.js]: For performance reason, it is better to move
`typeof costumeOpaqueSort === 'function' ? ...`
into setOpaqueSort/setTransparentSort functions.

File: src/renderers/webgl/WebGLRenderLists.js
Comment (donmccurdy) [inline src/renderers/webgl/WebGLRenderLists.js]: I think you mean "custom", not "costume", here?

Comment (ranbuch) [reply src/renderers/webgl/WebGLRenderLists.js]: Obviously, apologize.

File: src/renderers/webgl/WebGLRenderLists.js
Comment (ranbuch) [reply src/renderers/webgl/WebGLRenderLists.js]: Sure, good call!
I still have to check if `costumOpaqueSort` and `costumTransparentSort` are available for each sort iteration though.

**Comment by Mugen87** [comment]
@mrdoob The PR seems okay to me. If merged, I can add the docs and also update the TS file (unless @ranbuch does it in the meanwhile^^).

**Review by Mugen87 — APPROVED** [review]

**Comment by ranbuch** [comment]
> @mrdoob The PR seems okay to me. If merged, I can add the docs and also update the TS file (unless @ranbuch does it in the meanwhile^^).

Thank you @Mugen87 

Do you want me to provide the documentation? Should I use this feature branch to do so?
I suppose I'm should edit the `WebGLRenderer.html` file for the documentation.

About the `ts`, I see that there's a `WebGLRenderer.d.ts` file but it's probably auto generated. Where is the relevant `ts` source file?

**Comment by Mugen87** [comment]
>  see that there's a WebGLRenderer.d.ts file but it's probably auto generated.

No, it's manually maintained. 

> Should I use this feature branch to do so?

Yes, that's fine. Adding docs would be great.

**Comment by ranbuch** [comment]
@Mugen87 I don't mean to push it but is there an ETA?
I just delaying my viewer's update until the approval 😉

**Comment by Mugen87** [comment]
I'm afraid you have to wait for @mrdoob's review 😇

File: docs/api/en/renderers/WebGLRenderer.html
Comment (WestLangley) [inline docs/api/en/renderers/WebGLRenderer.html]: spelling: custom

many instances in this PR

File: src/renderers/WebGLRenderer.js
Comment (WestLangley) [inline src/renderers/WebGLRenderer.js]: Maybe it should be defined above (with _pixel ratio)
```js
_opaqueSort = null;
_transparentSort = null;
```
@mrdoob can decide.

File: src/renderers/WebGLRenderer.js
Comment (WestLangley) [inline src/renderers/WebGLRenderer.js]: We usually do not do type checking of args, but it is up to @mrdoob.

**Comment by ranbuch** [comment]
@WestLangley I've fixed all of the issues above on my local branch and can push if you want me to.

**Comment by Mugen87** [comment]
> and can push if you want me to.

I think that would be good 👍

**Comment by Mugen87** [comment]
Please don't push build files. It seems they sneaked in with the last commit.

**Comment by ranbuch** [comment]
@Mugen87 it seems I have conflicts, sorry about that.
The `Resolve conflicts` button is grayed out for me due to my write access.
Do you want me to do anything?

**Comment by Mugen87** [comment]
Do this on your system:
```
git reset HEAD~
```
Then undo the changes to the build files and make a new commit. (Force) push it like so:
```
git push origin feature/support-costume-painter-sort-functions-2 -f
```

File: docs/api/en/renderers/WebGLRenderer.html
Comment (WestLangley) [inline docs/api/en/renderers/WebGLRenderer.html]: >In some cases the transparent materials sort may get rendered in the wrong order. This issue may cause materials to render behind other materials instead of the front of them and vice versa.

I think that can be removed. File a bug report if you want.

Comment (ranbuch) [reply docs/api/en/renderers/WebGLRenderer.html]: @WestLangley I'm not sure I understand. I have opened a [bug report](https://github.com/mrdoob/three.js/issues/13889#issuecomment-385225320) and `mrdoob`'s solution was to add this function.

Please advise.

Comment (WestLangley) [reply docs/api/en/renderers/WebGLRenderer.html]: I am suggesting you remove those two sentences.

If this is merged, I think it would be a good idea to specify the calling arguments. See how `Obejct3D.onBeforeRender` is documented.

Comment (ranbuch) [reply docs/api/en/renderers/WebGLRenderer.html]: @WestLangley How about this:
`[method:null setTransparentSort]( [param:Function method] )`
`Sets the custom transparent sort function for the WebGLRenderLists. Pass null to use the default reversePainterSortStable function.`

Or:
`[method:null setTransparentSort]( [param:Function method] )`
`Sets the custom transparent sort function for the WebGLRenderLists. This function is called with the following parameter: method`

Or you can suggest anything else.

Comment (WestLangley) [reply docs/api/en/renderers/WebGLRenderer.html]: >If this is merged, I think it would be a good idea to specify the calling arguments. See how Obejct3D.onBeforeRender is documented.

Actually, what I was referring to was the signature of the function you are passing. Maybe @Mugen87 has a good suggestion for how to handle this...

Comment (Mugen87) [reply docs/api/en/renderers/WebGLRenderer.html]: I think a sentence like the following (from `Object3D.onBeforeRender()` is okay:
`This function is called with the following parameters: renderer, scene, camera, geometry, material, group.`

However, the TS declaration should be adjusted. Instead of using `Function`, it's better to  define the callback inline like so:
```ts
onBeforeRender: (
   renderer: WebGLRenderer,
   scene: Scene,
   camera: Camera,
   geometry: Geometry | BufferGeometry,
   material: Material,
   group: Group
) => void;
```

Comment (Mugen87) [reply docs/api/en/renderers/WebGLRenderer.html]: However, this can also be done with a separate PR. Maybe it's better to merge this one first and then move on.

Comment (WestLangley) [reply docs/api/en/renderers/WebGLRenderer.html]: Personally, I think the OP has a modeling problem. But @mrdoob thinks this is needed, then at least remove the two sentences I referenced prior to merging.

Comment (ranbuch) [reply docs/api/en/renderers/WebGLRenderer.html]: >How about this:
[method:null setTransparentSort]( [param:Function method] )
Sets the custom transparent sort function for the WebGLRenderLists. Pass null to use the default reversePainterSortStable function.
>Or:
[method:null setTransparentSort]( [param:Function method] )
Sets the custom transparent sort function for the WebGLRenderLists. This function is called with the following parameter: method
>Or you can suggest anything else.

@WestLangley Can you please write here the exact sentence you want me to push?

Comment (ranbuch) [reply docs/api/en/renderers/WebGLRenderer.html]: @Mugen87 can you please guide me on how to complete this PR?

Comment (Mugen87) [reply docs/api/en/renderers/WebGLRenderer.html]: The comment https://github.com/mrdoob/three.js/pull/16909/files#r352293599 asked you to remove the following two sentences.

> In some cases the transparent materials sort may get rendered in the wrong order. This issue may cause materials to rendere behind other materials instead of the front of them and vice versa.

Besides, I suggest update the TS declaration file. Instead of using `Function`, use this:
```js
onBeforeRender: (
   renderer: WebGLRenderer,
   scene: Scene,
   camera: Camera,
   geometry: Geometry | BufferGeometry,
   material: Material,
   group: Group
) => void;
```

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**
